### PR TITLE
Final bits of Player

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/globaldata.h [hunk 1]
```diff
@@ -192,12 +192,7 @@ class GlobalData : public SubsystemInterface
     Coord3D m_terrainLightPos[LIGHT_COUNT];
     float m_infantryLight[TIME_OF_DAY_COUNT];
     float m_infantryLightOverride; // Based on ScriptAction::doSetInfantryLightingOverride
-    float m_easySoloHumanHealthBonus;
-    float m_normalSoloHumanHealthBonus;
-    float m_hardSoloHumanHealthBonus;
-    float m_easySoloAIHealthBonus;
-    float m_normalSoloAIHealthBonus;
-    float m_hardSoloAIHealthBonus;
+    float m_soloHealthBonus[2][3];
```
Comment (xezon) [inline src/game/common/globaldata.h]: We need named values for this. With plain numbers this is unreadable.

File: src/game/logic/system/partitionmanager.h [hunk 2]
```diff
@@ -71,6 +72,24 @@ class PartitionFilterSameMapStatus : public PartitionFilter
...
 
+class PartitionFilterAcceptByKindOf : public PartitionFilter
+{
+public:
+    PartitionFilterAcceptByKindOf(BitFlags<KINDOF_COUNT> &must_be_set, BitFlags<KINDOF_COUNT> &must_be_clear) :
+        m_mustBeSet(must_be_set), m_mustBeClear(must_be_clear)
+    {
+    }
+
+    virtual bool Allow(Object *obj);
```
Comment (xezon) [inline src/game/logic/system/partitionmanager.h]: override

File: src/game/logic/object/object.cpp [hunk 3]
```diff
@@ -732,3 +732,49 @@ bool Object::Set_Script_Status(ObjectScriptStatusBit bit, bool set)
...
+
+        if (update != nullptr) {
+            return update;
+        }
+    }
+
+    return nullptr;
+}
+
+void Object::Set_Vision_Spied(bool vision_spied, int player)
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: player_index

File: src/game/logic/object/object.cpp [hunk 4]
```diff
@@ -732,3 +732,49 @@ bool Object::Set_Script_Status(ObjectScriptStatusBit bit, bool set)
...
+{
+    bool changed = false;
```
Comment (xezon) [inline src/game/logic/object/object.cpp:756]: No assert here to verify player index bounds?

File: src/game/logic/object/object.cpp [hunk 5]
```diff
@@ -732,3 +732,49 @@ bool Object::Set_Script_Status(ObjectScriptStatusBit bit, bool set)
...
+
+    if (vision_spied) {
+        if (++m_spiedOnByPlayer[player] == 1) {
+            changed = 1;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: true

File: src/game/logic/object/object.cpp [hunk 6]
```diff
@@ -732,3 +732,49 @@ bool Object::Set_Script_Status(ObjectScriptStatusBit bit, bool set)
...
+        }
+    } else if (!--m_spiedOnByPlayer[player]) {
+        changed = 1;
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: true

File: src/game/logic/object/object.cpp [hunk 7]
```diff
@@ -732,3 +732,49 @@ bool Object::Set_Script_Status(ObjectScriptStatusBit bit, bool set)
...
+
+void Object::Set_Vision_Spied(bool vision_spied, int player)
+{
+    bool changed = false;
+
+    if (vision_spied) {
+        if (++m_spiedOnByPlayer[player] == 1) {
+            changed = 1;
+        }
+    } else if (!--m_spiedOnByPlayer[player]) {
```
Comment (xezon) [inline src/game/logic/object/object.cpp]: == 0

File: src/game/logic/object/object.cpp [hunk 8]
```diff
@@ -732,3 +732,49 @@ bool Object::Set_Script_Status(ObjectScriptStatusBit bit, bool set)
...
+        changed = 1;
+    }
+
+    if (changed) {
+        unsigned short flags = 0;
```
Comment (xezon) [inline src/game/logic/object/object.cpp:767]:
```c++
static_assert(sizeof(flags) * 8 <= MAX_PLAYER_COUNT, "Bit field must be large enough for player count")
```
please

File: src/game/logic/object/update/aiupdate.h [hunk 9]
```diff
@@ -176,6 +193,27 @@ class AICommandInterface
...
     }
+
+    void AI_Enter(Object *obj, CommandSourceType cmd_source)
+    {
+        AICommandParms params(AICMD_ENTER, cmd_source);
+        params.m_obj = obj;
+        AI_Do_Command(&params);
+    }
+
+    void AI_Evacuate(bool b, CommandSourceType cmd_source)
```
Comment (xezon) [inline src/game/logic/object/update/aiupdate.h]: bool for_attack
or
bool reenter

The AssaultTransport has this feature where transported units will temporarily evacuate to attack a target, and once the attack is completed, they will automatically reenter the vehicle.

File: src/game/logic/ai/squad.cpp [hunk 10]
```diff
@@ -48,3 +52,113 @@ bool Squad::Is_On_Squad(const Object *obj) const
...
+        }
+    }
+}
+
+std::vector<Object *> &Squad::Get_All_Objects()
+{
+    m_objectsCached.clear();
+
+    for (auto i = m_objectIDs.begin(); i != m_objectIDs.end();) {
+        Object *obj = g_theGameLogic->Find_Object_By_ID((*i));
```
Comment (xezon) [inline src/game/logic/ai/squad.cpp]: Can remove obsolete ()

File: src/game/logic/ai/squad.cpp [hunk 11]
```diff
@@ -48,3 +52,113 @@ bool Squad::Is_On_Squad(const Object *obj) const
...
+            m_objectIDs.erase(it);
+        }
+    }
+}
+
+std::vector<Object *> &Squad::Get_All_Objects()
+{
+    m_objectsCached.clear();
+
+    for (auto i = m_objectIDs.begin(); i != m_objectIDs.end();) {
```
Comment (xezon) [inline src/game/logic/ai/squad.cpp]: it

File: src/game/logic/ai/squad.cpp [hunk 12]
```diff
@@ -48,3 +52,113 @@ bool Squad::Is_On_Squad(const Object *obj) const
...
+    }
+
+    return m_objectsCached;
+}
+
+std::vector<Object *> &Squad::Get_Live_Objects()
+{
+    Get_All_Objects();
+
+    for (auto i = m_objectsCached.begin(); i != m_objectsCached.end();) {
```
Comment (xezon) [inline src/game/logic/ai/squad.cpp]: it

File: src/game/logic/ai/squad.cpp [hunk 13]
```diff
@@ -48,3 +52,113 @@ bool Squad::Is_On_Squad(const Object *obj) const
...
+            i = m_objectsCached.erase(i);
+        }
+    }
+
+    return m_objectsCached;
+}
+
+int Squad::Get_Size_Of_Group() const
+{
+    return m_objectIDs.size();
```
Comment (xezon) [inline src/game/logic/ai/squad.cpp]: `static_cast<int>`

File: src/game/logic/ai/squad.cpp [hunk 14]
```diff
@@ -48,3 +52,113 @@ bool Squad::Is_On_Squad(const Object *obj) const
...
+        m_objectIDs = from_ai_group->Get_All_IDs();
+    }
+}
+
+void Squad::AI_Group_From_Squad(AIGroup *ai_group_to_fill)
+{
+    if (ai_group_to_fill != nullptr) {
+        Get_Live_Objects();
+
+        for (auto i = m_objectsCached.begin(); i != m_objectsCached.end(); i++) {
```
Comment (xezon) [inline src/game/logic/ai/squad.cpp]: it

**Review by xezon — COMMENTED** [review]
Review progress so far. A few files left. I will continue tomorrow.

File: src/game/logic/system/partitionmanager.h (hunk 2 above)
Comment (jonwil) [reply src/game/logic/system/partitionmanager.h]: fixed.

File: src/game/logic/object/object.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp:756]: nope.

File: src/game/logic/object/object.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/object.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/object.cpp]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

File: src/game/logic/ai/squad.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/ai/squad.cpp]: fixed.

File: src/game/logic/ai/squad.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/logic/ai/squad.cpp]: fixed.

File: src/game/logic/ai/squad.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/ai/squad.cpp]: fixed.

File: src/game/logic/ai/squad.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/logic/ai/squad.cpp]: fixed.

File: src/game/logic/ai/squad.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/logic/ai/squad.cpp]: fixed.

File: src/game/common/globaldata.h (hunk 1 above)
Comment (jonwil) [reply src/game/common/globaldata.h]: After some more thinking and some discussion with toms, I came up with something for this.

File: src/game/common/globaldata.cpp [hunk 15]
```diff
@@ -974,12 +980,12 @@ GlobalData::GlobalData()
...
-    m_hardSoloHumanHealthBonus = 1.0f;
-    m_easySoloAIHealthBonus = 1.0f;
-    m_normalSoloAIHealthBonus = 1.0f;
-    m_hardSoloAIHealthBonus = 1.0f;
+    m_soloHealthBonus[0][0] = 1.0f;
+    m_soloHealthBonus[0][1] = 1.0f;
+    m_soloHealthBonus[0][2] = 1.0f;
+    m_soloHealthBonus[1][0] = 1.0f;
+    m_soloHealthBonus[1][1] = 1.0f;
+    m_soloHealthBonus[1][2] = 1.0f;
```
Comment (xezon) [inline src/game/common/globaldata.cpp]: named indexing missing

File: src/game/common/rts/player.cpp [hunk 16]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+
+                    if (obj != nullptr) {
+                        AIUpdateInterface *update = obj->Get_AI_Update_Interface();
+
+                        if (update != nullptr) {
+                            for (Object *o = iter->First(); o != nullptr; o = iter->Next()) {
+                                ContainModuleInterface *contain = o->Get_Contain();
+
+                                if (contain != nullptr) {
+                                    unsigned short id = contain->Get_Player_Who_Entered();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: mask

File: src/game/common/rts/player.cpp [hunk 17]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+
+    while (m_currentSkillPoints >= m_skillPointsNeededForNextRank) {
+        Set_Rank_Level(m_rankLevel + 1);
+        ret = true;
+    }
+
+    return ret;
+}
+
+bool Player::Set_Rank_Level(int rank)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:3039]: new_rank

File: src/game/common/rts/player.cpp [hunk 18]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+        return false;
+    }
+
+    int points = m_sciencePurchasePoints;
+
+    if (rank < m_rankLevel) {
+        Reset_Rank();
+    }
+
+    for (int i = m_rankLevel + 1; i <= rank; i++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:3063]: r

File: src/game/common/rts/player.cpp [hunk 19]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+
+            if (m_sciencePurchasePoints < 0) {
+                m_sciencePurchasePoints = 0;
+            }
+
+            if (m_currentSkillPoints < info->m_skillPointsNeeded) {
+                m_currentSkillPoints = info->m_skillPointsNeeded;
+            }
+
+            for (auto j = info->m_sciencesGranted.begin(); j != info->m_sciencesGranted.end(); j++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: science_it

File: src/game/common/rts/player.cpp [hunk 20]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+    return true;
+}
+
+bool Player::Add_Skill_Points_For_Kill(const Object *killer, const Object *killed)
+{
+    if (killer != nullptr && killed != nullptr && !killed->Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION)) {
+        return Add_Skill_Points(killed->Get_Template()->Get_Skill_Point_Value(killed->Get_Veterancy_Level()));
+    }
+
+    return 0;
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: false

File: src/game/common/rts/player.cpp [hunk 21]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+        skill = 0x7FFFFFFF;
+    }
+
+    m_skillPointsNeededForNextRank = skill;
+    m_rankLevel = rank;
+    captainslog_dbgassert(
+        m_currentSkillPoints >= m_rankProgress && m_currentSkillPoints < m_skillPointsNeededForNextRank, "hmm, wrong");
+
+    if (g_theControlBar != nullptr) {
+        if (m_skillPointsNeededForNextRank && Is_Local_Player()) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: != 0

File: src/game/common/rts/player.cpp [hunk 22]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+bool Player::Can_Build_More_Of_Type(const ThingTemplate *tmplate) const
+{
+    int max = tmplate->Get_Max_Simultaneous_Of_Type();
+
+    if (max != 0) {
+        CanBuildStruct can_build;
+        can_build.m_count = 0;
+        can_build.m_thingTemplate = tmplate;
+        can_build.m_linkKey = tmplate->Get_Max_Simultaneous_Link_Key();
+        can_build.m_isStructure = !tmplate->Is_KindOf(KINDOF_STRUCTURE);
```
Comment (xezon) [inline src/game/common/rts/player.cpp:3198]: m_isStructure is not KINDOF_STRUCTURE ?

File: src/game/common/rts/player.cpp [hunk 23]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+    }
+
+    if (tmplate->Get_Buildable() == BSTATUS_ONLY_AI && Get_Player_Type() != PLAYER_COMPUTER) {
+        return false;
+    }
+
+    bool ret = true;
+
+    for (int i = 0; i < tmplate->Get_Prereq_Count(); i++) {
+        if (!tmplate->Get_Nth_Prereq(i)->Is_Satisifed(this)) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp:3234]: Is_Satisfied type error

File: src/game/common/rts/player.cpp [hunk 24]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+        ret = true;
+    }
+#endif
+
+    return ret && Can_Build_More_Of_Type(tmplate);
+}
+
+void Player::Friend_Apply_Difficulty_Bonuses_For_Object(Object *obj, bool apply)
+{
+    static WeaponBonusConditionType wbonus[2][3] = { { WEAPONBONUSCONDITION_SOLO_HUMAN_EASY,
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: Named indexing missing

File: src/game/common/rts/player.cpp [hunk 25]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+    }
+
+    if (!added) {
+        if (!removed) {
+            return;
+        }
+
+        float armor;
+
+        if (bonuses->m_armorBonus <= 0.01) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: f

File: src/game/common/rts/player.cpp [hunk 26]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+            armor = 0.01f;
+        } else {
+            armor = bonuses->m_armorBonus;
+        }
+
+        bonuses->m_armorBonus = 1.0f / armor;
+        float sight;
+
+        if (bonuses->m_sightBonus <= 0.01) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: f

File: src/game/common/rts/player.cpp [hunk 27]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+    }
+}
+
+void Player::Process_Select_Team_GameMessage(int squad, GameMessage *message)
+{
+    if (squad < SQUAD_COUNT) {
+        if (m_squads[squad] != nullptr) {
+            m_squads[squad]->Clear_Squad();
+            std::vector<Object *> vector(m_squads[squad]->Get_Live_Objects());
+            int size = vector.size();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: `static_cast` or `size_t`

File: src/game/common/rts/player.cpp [hunk 28]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+void Player::Process_Add_Team_GameMessage(int squad, GameMessage *message)
+{
+    if (squad < SQUAD_COUNT) {
+        if (m_squads[squad] != nullptr) {
+            if (m_aiSquad == nullptr) {
+                m_aiSquad = new Squad();
+            }
+
+            std::vector<Object *> vector(m_squads[squad]->Get_Live_Objects());
+            int size = vector.size();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: `static_cast` or `size_t`

File: src/game/common/rts/player.cpp [hunk 29]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+    }
+}
+
+void Player::Set_Currently_Selected_AIGroup(AIGroup *group)
+{
+    if (m_aiSquad == nullptr) {
+        m_aiSquad = new Squad();
+    }
+
+    m_aiSquad->Clear_Squad();
```
Comment (xezon) [inline src/game/common/rts/player.cpp:3483]: Could put this in else

File: src/game/common/rts/player.cpp [hunk 30]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+}
+
+void Player::Add_AIGroup_To_Current_Selection(AIGroup *group)
+{
+    if (m_aiSquad == nullptr) {
+        m_aiSquad = new Squad();
+    }
+
+    std::vector<ObjectID> vector(group->Get_All_IDs());
+    int size = vector.size();
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: `static_cast` or `size_t`

File: src/game/common/rts/player.cpp [hunk 31]
```diff
@@ -2802,3 +2806,732 @@ unsigned int Player::Get_Or_Start_Special_Power_Ready_Frame(const SpecialPowerTe
...
+        }
+    }
+}
+
+void Player::Set_Units_Vision_Spied(bool vision_spied, BitFlags<KINDOF_COUNT> flags, int player)
+{
+    VisionSpiedStruct s;
+    s.vision_spied = vision_spied;
+    s.flags = flags;
+    s.player = player;
```
Comment (xezon) [inline src/game/common/rts/player.cpp:3536]: This function is doing nothing.

File: src/game/common/globaldata.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/common/globaldata.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:3198]: Yes, this is correct.

File: src/game/common/rts/player.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 25 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 26 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 27 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 28 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 30 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 31 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:3536]: fixed.

**Review by xezon — COMMENTED** [review]
Teams select is broken. Upon select, the team is removed.

CTRL + Key 1, assigns team
Key 1, selects team and removes team 

shot_20230306_224534_1

**Review by xezon — COMMENTED** [review]
When ranked up, the EVA sound is missing for it.

shot_20230306_225647_5

**Comment by xezon** [comment]
Battle Plan works.
Cash Bounty works.

**Comment by jonwil** [comment]
Fixed the issue with team selection.
As for the other issue, I tested with "USA Laser General" on map "Alpine Assault" against an easy army and when it ranked up I heard the "congratulations general you have been promoted" sound as expected.

**Comment by xezon** [comment]
Team selection issue is fixed.
~~Level up EVA issue is fixed.~~

**Comment by xezon** [comment]
I do not know how to test

Player::Transfer_Assets_From_That
Player::Garrison_All_Units
Player::Ungarrison_All_Units
Player::Set_Units_Should_Idle_Or_Resume
Player::Sell_Everything_Under_The_Sun

**Comment by xezon** [comment]
Max Simultaneous works. Tested with USA Strategy Center, China Internet Center, China Hero unit.

**Comment by xezon** [comment]
Correction: Level up EVA issue is NOT fixed. Reproduced in Debug configuration on multiple maps. Checked against Original.

https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Maps/SkirmishOnly/!SkirmishTestPromotionsCHINA
https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Maps/SkirmishOnly/!SkirmishTestPromotionsGLA
https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Maps/SkirmishOnly/!SkirmishTestPromotionsUSA
https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Maps/Any/!TestRank

**Comment by xezon** [comment]
There is an issue with Cash Bounty reward value. It looks like the value is always 1 cash higher than it should be.

## Original

shot_20230307_175918_1

## Thyme

shot_20230307_175711_2

**Comment by jonwil** [comment]
Tested many maps and many tests and I get the "congratulations general you have been promoted" sound every time.

**Review by xezon — APPROVED** [review]
Looks ok. Unable to test some of the code because it appears to be special script functionality. There are some issues with Cash Bounty and EVA sounds, which are unrelated to this change and are now tracked with
#883, #884

Tested in Debug and RelWithDebInfo. No crashes.

**Pull request merged**